contracts/contracts/strategies/GeneralizedUniswapV3Strategy.sol

```diff
@@ -0,0 +1,850 @@
+        if (token1Balance < minAmount1) {
+            vault.withdrawForUniswapV3(
+                address(this),
+                token1,
+                minAmount1 - token1Balance
+            );
+        }
+    }
+
+    /***************************************
+            Balances and Fees
+    ****************************************/
+
+    /**
+     * @notice Collect accumulated fees from the active position
+     * @dev Doesn't send to vault or harvester
+     */
+    function collectRewardTokens()
+        external
+        override
+        onlyHarvester
+        nonReentrant
+    {
+        if (currentPositionTokenId > 0) {
+            _collectFeesForToken(currentPositionTokenId);
+        }
+    }
+
+    /**
+     * @notice Returns the accumulated fees from the active position
+     * @return amount0 Amount of token0 ready to be collected as fee
+     * @return amount1 Amount of token1 ready to be collected as fee
+     */
+    function getPendingFees()
+        external
+        view
+        returns (uint256 amount0, uint256 amount1)
+    {
+        (amount0, amount1) = uniswapV3Helper.positionFees(
+            positionManager,
+            platformAddress,
+            currentPositionTokenId
+        );
+    }
+
+    /**
+     * @notice Collects the fees generated by the position on V3 pool
+     * @param tokenId Token ID of the position to collect fees of.
+     * @return amount0 Amount of token0 collected as fee
+     * @return amount1 Amount of token1 collected as fee
+     */
+    function _collectFeesForToken(uint256 tokenId)
+        internal
+        returns (uint256 amount0, uint256 amount1)
+    {
+        INonfungiblePositionManager.CollectParams
+            memory params = INonfungiblePositionManager.CollectParams({
+                tokenId: tokenId,
+                recipient: address(this),
+                amount0Max: type(uint128).max,
+                amount1Max: type(uint128).max
+            });
+
+        (amount0, amount1) = positionManager.collect(params);
+
+        emit UniswapV3FeeCollected(tokenId, amount0, amount1);
+    }
+
+    /**
+     * @dev Only checks the active LP position.
+     *      Doesn't return the balance held in the reserve strategies.
+     * @inheritdoc InitializableAbstractStrategy
+     */
+    function checkBalance(address _asset)
+        external
+        view
+        override
+        returns (uint256 balance)
+    {
+        require(_asset == token0 || _asset == token1, "Unsupported asset");
+
+        balance = IERC20(_asset).balanceOf(address(this));
+
+        (uint160 sqrtRatioX96, , , , , , ) = IUniswapV3Pool(platformAddress)
+            .slot0();
+
+        if (currentPositionTokenId > 0) {
+            require(
+                tokenIdToPosition[currentPositionTokenId].exists,
+                "Invalid token"
+            );
+
+            (uint256 amount0, uint256 amount1) = uniswapV3Helper.positionValue(
```

Comment:
Calculating a strategies balance is the place where gas efficiency matters most. We go through every coin, in every strategy, every rebase, redeem, large mint. And we do it at least twice during allocations. 

Given that the operator is probably rebalancing every day or few, it's might be worth considering how much extra gas taking into account fees in this calculation will cost us.